### PR TITLE
:sparkles: Kai support

### LINKED ISSUE #630: :sparkles: Support AI Service
### The AI Service needs ###

To synchronize analysis reports. This requires:
- The archiving feature to be adjusted. Reports cannot be archived until KAI has retrieved them.
- List analysis reports with filtering by id.

#### Fetch Analysis Reports ####

Add `/analyses` endpoint with filtering on `id`.  Consistent with `/applications/analyses`  returns a _sparse_ report (empty issues).  As a policy, the hub cannot hold an Analysis report in memory because each report can be very large.  The is to prevent multiple/concurrent API requests from exceeding 2G memory limit.  For this reason, endpoints cannot return arrays or large resources either. 

Example:  _assumes URL encoding_
```
/analyses?filter=id>45 
```

A stretch goal would be to add a kubernetes (like) watch mechanism (like we did in MTV). 

Perhaps:
```
GET /watch/analyses
GET /watch/analyses?filter=id>18
```

For each resource returned, fetch the complete report using: `/analyses/:id`.

#### Retention Policy ####

Add a setting to define the analysis report retention policy.  This will determine when an analysis report may be _archived_.  Thinking the policy should be expressed as the number of un-archived reports for each application (beyond the latest). Values:
- 0 (default) - previous reports archived immediately.
- 1+ - number of previous ports to remain un-archived.
- -1 - unlimited

A new endpoint will need to be added to archive a report (after synchronized by KAI).
Perhaps:  POST /analyses/:id/archive

#### Commit (hash) ####

Add a field to the `Analysis` resource.  In an effort to support both Git and Subversion, the filed would be named: `commit`.  For git this would be a hash.

The addon-analyzer will need to set this field.

### PR DESCRIPTION
**Pull request by jortel**
closes #630 

- Added endpoint:
  - GET /applications/:id/analyses/issues
  - GET /analyses
  - GET /analyses/:id/issues
- Updated api.Analysis to omit issues, summary. Struct never used this way.
- Add field: Analysis.Commit. migration=13.

Added IssueWriter to support new endpoints.

---
migration/14
```
diff -bur migration/v13/model/analysis.go migration/v14/model/analysis.go
--- migration/v13/model/analysis.go	2024-06-11 03:29:34.402853992 -0700
+++ migration/v14/model/analysis.go	2024-06-11 04:29:29.898613621 -0700
@@ -6,7 +6,8 @@
 type Analysis struct {
 	Model
 	Effort        int
-	Archived      bool             `json:"archived"`
+	Commit        string
+	Archived      bool
 	Summary       JSON             `gorm:"type:json"`
 	Issues        []Issue          `gorm:"constraint:OnDelete:CASCADE"`
 	Dependencies  []TechDependency `gorm:"constraint:OnDelete:CASCADE"`
diff -bur migration/v13/model/application.go migration/v14/model/application.go
--- migration/v13/model/application.go	2024-06-11 03:29:34.402853992 -0700
+++ migration/v14/model/application.go	2024-06-11 04:29:29.898613621 -0700
@@ -76,7 +76,7 @@
 	return
 }
 
-// Validation Hook to avoid cyclic dependencies.
+// BeforeCreate detects cyclic dependencies.
 func (r *Dependency) BeforeCreate(db *gorm.DB) (err error) {
 	var nextDeps []*Dependency
 	var nextAppsIDs []uint
@@ -96,11 +96,11 @@
 	return
 }
 
-// Custom error type to allow API recognize Cyclic Dependency error and assign proper status code.
+// DependencyCyclicError reports cyclic Dependency error.
 type DependencyCyclicError struct{}
 
-func (err DependencyCyclicError) Error() string {
-	return "cyclic dependencies are not allowed"
+func (e DependencyCyclicError) Error() string {
+	return "Cyclic dependencies are not permitted."
 }
 
 type BusinessService struct {

```

### CODE REVIEW

File: migration/v13/model/analysis.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+import (
+	"gorm.io/gorm"
+)
+
+// Analysis report.
+type Analysis struct {
+	Model
+	Effort        int
+	Commit        string           `json:"commit"`
```
Comment (jortel) [inline migration/v13/model/analysis.go:11]: added Commit.

File: api/analysis.go [hunk 2]
```diff
@@ -282,6 +366,16 @@ func (h AnalysisHandler) AppCreate(ctx *gin.Context) {
...
+	analysis := r.Model()
+	analysis.ApplicationID = id
+	analysis.CreateUser = h.BaseHandler.CurrentUser(ctx)
+	db := h.DB(ctx)
+	db.Logger = db.Logger.LogMode(logger.Error)
+	err = db.Create(analysis).Error
+	if err != nil {
+		_ = ctx.Error(err)
+		return
+	}
```
Comment (jortel) [inline api/analysis.go:378]: hint: Moved this code blocker here to use r.Model() to properly transfer analysis fields in the posted resource (like: Commit).

File: migration/v14/model/analysis.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
+package model
+
+import "gorm.io/gorm"
+
+// Analysis report.
+type Analysis struct {
+	Model
+	Effort        int
+	Commit        string
```
Comment (jortel) [inline migration/v14/model/analysis.go:9]: **hint**: adds commit.

File: api/analysis.go [hunk 4]
```diff
@@ -110,6 +120,64 @@ func (h AnalysisHandler) Get(ctx *gin.Context) {
...
 }
 
+// List godoc
+// @summary List analyses.
+// @description List analyses.
+// @description Resources do not include relations.
+// @tags analyses
+// @produce json
+// @success 200 {object} []api.Analysis
+// @router /analyses [get]
```
Comment (jortel) [inline api/analysis.go:130]: **hint**: added for completeness.

File: api/analysis.go [hunk 5]
```diff
@@ -141,6 +209,32 @@ func (h AnalysisHandler) AppLatest(ctx *gin.Context) {
...
 }
 
+// Archive godoc
+// @summary Archive an analysis (report) by ID.
+// @description Archive an analysis (report) by ID.
+// @tags analyses
+// @produce octet-stream
+// @success 204 {object}
+// @router /analyses/{id}/archive [post]
+// @param id path int true "Analysis ID"
```
Comment (jortel) [inline api/analysis.go:219]: **hint**: added to support manual archiving when automatic archiver disabled.

File: api/analysis.go [hunk 6]
```diff
@@ -185,6 +279,8 @@ func (h AnalysisHandler) AppList(ctx *gin.Context) {
 	db := h.DB(ctx)
 	db = db.Model(&model.Analysis{})
 	db = db.Where("ApplicationID = ?", id)
+	db = db.Preload("Application")
```
Comment (jortel) [inline api/analysis.go:282]: **hint**: preload application so name is set in ref.

File: api/analysis.go [hunk 7]
```diff
@@ -388,7 +482,7 @@ func (h AnalysisHandler) AppCreate(ctx *gin.Context) {
 	}
 
 	db = h.DB(ctx)
-	db = db.Preload(clause.Associations)
+	db = db.Preload("Application")
```
Comment (jortel) [inline api/analysis.go:485]: **hint**: only preload app.  Don't need/want issues.

File: api/analysis.go [hunk 8]
```diff
@@ -626,52 +692,73 @@ func (h AnalysisHandler) Issues(ctx *gin.Context) {
...
+// @description - rule
+// @description - name
+// @description - category
+// @description - effort
+// @description - labels
+// @tags issues
+// @produce json
+// @success 200 {object} []api.Issue
+// @router /analyses/{id}/issues [get]
+// @param id path int true "Analysis ID"
```
Comment (jortel) [inline api/analysis.go:728]: **hint**: needed by kai to fetch issues.

File: migration/v14/migrate.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package v14
+
+import (
+	"github.com/jortel/go-utils/logr"
+	"github.com/konveyor/tackle2-hub/migration/v14/model"
+	"gorm.io/gorm"
+)
+
+var log = logr.WithName("migration|v13")
```
Comment (mansam) [inline migration/v14/migrate.go:9]: log message needs to be updated to v14

**Review by mansam — APPROVED** [review]
LGTM, just little nits

**Pull request merged**